src/PerfTreeParser/PerfTreeParser.class.st

```diff
@@ -0,0 +1,244 @@
+
+	| fileContents i |
+	fileContents := aFile contents lines.
+	i := 1.
+
+	[ (fileContents at: i) beginsWith: '# sample duration' ] whileFalse: [
+		i := i + 1 ].
+
+	self time: (((fileContents at: i) last: 10) first: 7) asNumber.
+
+	content := fileContents reject: [ :line | line beginsWith: '#' ].
+	content := content select: [ :line |
+		           ((line copyWithout: Character space) copyWithout: $|)
+			           isNotEmpty ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> differenceOfSpacesAt: aNodeIndex [
+	"Gives the difference of the number of spaces between 2 followed lines."
+
+	^ (aNodeIndex between: 1 and: self numberOfFunctions)
+		  ifFalse: [ nil ]
+		  ifTrue: [
+			  (self numberOfSpacesAt: aNodeIndex + 1)
+			  - (self numberOfSpacesAt: aNodeIndex) ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> findChildrenIndexOfIndex: aNode [
+	"Gives an ordered collection of every indexes of the lines where there is a child."
+
+	| lastNode everyChildren |
+	lastNode := self numberOfFunctions.
+	everyChildren := OrderedCollection new.
+
+	(aNode between: 1 and: lastNode) ifFalse: [ ^ 0 ].
+
+	(((self differenceOfSpacesAt: aNode) between: 1 and: 10) or: [
+		 (self percentageOf: aNode) = (self percentageOf: aNode + 1) ])
+		ifTrue: [
+			everyChildren add: aNode + 1.
+			^ everyChildren ].
+
+	aNode to: lastNode do: [ :i |
+		(self numberOfSpacesAt: i) - (self numberOfSpacesAt: aNode) = 11
+			ifTrue: [ everyChildren add: i ] ].
+
+	^ everyChildren
+]
+
+{ #category : #accessing }
+PerfTreeParser >> findChildrenOfIndex: aNodeIndex [
+	"Gives the nodes representing every children at the node aNodeIndex."
+
+	| collectionOfChildrenIndexes |
+	collectionOfChildrenIndexes := self findChildrenIndexOfIndex: aNodeIndex.
+	^ collectionOfChildrenIndexes collect: [ :childIndex |
+		  self nodeAtIndex: childIndex ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> lastSpaceAt: aLine [
+	"Gives the index of the last space or '-' at a given line."
+
+	| lastSpace line |
+	(aLine between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
+
+	line := self readLine: aLine.
+	lastSpace := 0.
+
+	1 to: line size do: [ :i |
+		((line at: i) = Character space or: [ (line at: i) = $- ]) ifTrue: [
+			lastSpace := i ] ].
+	^ lastSpace
+]
+
+{ #category : #accessing }
+PerfTreeParser >> nameOf: aNodeIndex [
+	"Gives the name of a function at aNodeIndex."
+
+	| aLine |
+	(aNodeIndex between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
+
+	aLine := self readLine: aNodeIndex.
+	^ aLine last: aLine size - (self lastSpaceAt: aNodeIndex)
+]
+
+{ #category : #accessing }
+PerfTreeParser >> nodeAtIndex: aNodeIndex [
+	"Gives the node representing the node at aNodeIndex."
+
+	| child |
+	(aNodeIndex between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
+
+	child := PerfTreeNode new
+		         name: (self nameOf: aNodeIndex);
+		         weight: (self weightOf: aNodeIndex);
+		         time: self time;
+		         yourself.
```

Comment:
This is the `child` or the `node`?

Reply:
It's the `node`, before the classes had misleading names and this is something I should have changed too.